core/admin/mailu/utils.py

```diff
@@ -48,3 +71,359 @@ def init_app(self, app):
+                pass
+            else:
+                (self._uid, self._sid, self._created) = parsed
+                self._key = key
+
+        if initial is None:
+            # start new session
+            self.new = True
+            self._uid = None
+            self._sid = None
+            self._created = self.app.session_config.gen_created()
+            self._key = None
+
+        def _on_update(obj):
+            obj.modified = True
+
+        CallbackDict.__init__(self, initial, _on_update)
+
+    @property
+    def saved(self):
+        """ this reflects if the session was saved. """
+        return self._key is not None
+
+    @property
+    def sid(self):
+        """ this reflects the session's id. """
+        if self._sid is None or self._uid is None or self._created is None:
+            return None
+        return b''.join([self._uid, self._sid, self._created])
+
+    def destroy(self):
+        """ destroy session for security reasons. """
+
+        self.delete()
+
+        self._uid = None
+        self._sid = None
+        self._created = None
+
+        self.clear()
+
+        self.modified = True
+        self.new = False
+
+    def regenerate(self):
+        """ generate new id for session to avoid `session fixation`. """
+
+        self.delete()
+
+        self._sid = None
+        self._created = self.app.session_config.gen_created()
+
+        self.modified = True
+
+    def delete(self):
+        """ Delete stored session. """
+        if self.saved:
+            self.app.session_store.delete(self._key)
+            self._key = None
+
+    def save(self):
+        """ Save session to store. """
+
+        # set uid from dict data
+        if self._uid is None:
+            self._uid = self.app.session_config.gen_uid(self.get('user_id', ''))
+
+        # create new session id for new or regenerated sessions
+        if self._sid is None:
+            self._sid = self.app.session_config.gen_sid()
+
+        # get new session key
+        key = self.sid
+
+        # delete old session if key has changed
+        if key != self._key:
+            self.delete()
+
+        # save session
+        self.app.session_store.put(
+            key,
+            pickle.dumps(dict(self)),
+            self.app.permanent_session_lifetime.total_seconds()
+        )
+
+        self._key = key
+
+        self.new = False
+        self.modified = False
+
+class MailuSessionConfig:
+    """ Stores sessions crypto config """
+
+    def __init__(self, app=None):
+
+        if app is None:
+            app = flask.current_app
+
+        bits = app.config.get('SESSION_KEY_BITS', 64)
+
+        if bits < 64:
+            raise ValueError('Session id entropy must not be less than 64 bits!')
+
+        hash_bytes = bits//8 + (bits%8>0)
+        time_bytes = 4 # 32 bit timestamp for now
+        shaker = hashlib.shake_256 if bits>128 else hashlib.shake_128
+
+        self._shaker   = shaker(want_bytes(app.config.get('SECRET_KEY', '')))
```

Comment:
what do you suggest? note that the secret key is just used as a seed for the shaker here.
gen_uid() uses this seeded shaker later in line 277-279
I could use pbkds2_hmac, but I can't use a random seed for this:

```
skey = app.config.get('SECRET_KEY', '')
self._shaker = hashlib.shake_128(hashlib.pbkdf2_hmac('sha256', skey, skey, 100000))
```

Reply:
Just derive a specific key using hmac(SECRET_KEY, "cookie_username_key") ... like what's done in https://github.com/Mailu/Mailu/pull/1754/files

The goal here is to ensure there are enough bytes to seed the shaker (regardless of SECRET_KEY's length). There is no need to iterate over it IMHO.

Why are we keying it in the first place instead of doing H(email) ? Was it to ensure that the sessions are invalidated if SECRET_KEY changes or something else?

Reply:
If that was the goal maybe we should store that hash in the session itself and upon session validation decide on whether to accept (or not) the session

Reply:
I think it was my idea to use SECRET_KEY as a seed here - for no particular reason. At least I don't remember.
I could remove the shaker and replace gen_uid with something like this:

```
def gen_uid(self, uid):
  """ Generate hashed user id part of session key. """
  return hmac.digest(app.secret_key, want_bytes(uid), digest='sha256')[:self._hash_len]
```

Reply:
I would either:
- not key it at all and use the hash of the uid
- use hmac: derive a specific key from secret_key, use that derived key to hmac against the uid 

and truncate the output to what we deem to be appropriate in the session_id.

Thinking about it twice, maybe that's what we should make configurable: how much to truncate it (as that depends on the total number of users expected on the system) instead of how many bits to use in the random part (128bits should be enough for everyone for the foreseeable future)

Reply:
I've implemented option 2 (hmac of derived app secret + uid) truncated to SESSION_KEY_BITS.
The random part is fixed to 128 bits now.

I've also added the redis-ttl-refresh, as I did not observe the csrf_token changing.